src/rt/lifetime.d

```diff
@@ -281,15 +283,7 @@ bool __setArrayAllocLength(ref BlkInfo info, size_t newlength, bool isshared, co
         {
             if(isshared)
             {
-                try synchronized(typeid(ArrayAllocLengthLock))
-                {
-                    if(*length == cast(ubyte)oldlength)
-                        *length = cast(ubyte)newlength;
-                    else
-                        return false;
-                }
-                catch (Throwable t)
-                    assert(0, "Failed to synchronize.");
+                return cas(cast(shared)length, cast(ubyte)oldlength, cast(ubyte)newlength);
```

Comment:
> I don't agree that locking a mutex is any different than touching shared data.

Once again shared is a type qualifier while pure ensures that you don't access global state, those are distinct concepts even though they often overlap.

This is perfectly pure.
```d
void foo() pure
{
    auto ary = new shared(int[])(10);
    ary.length = 20;
}
```

A mutex can change global state in the kernel which is a side-effect and thus not pure.